src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py

```diff
@@ -176,61 +207,214 @@ def __blob_is_not_complete(metadata):
+        for name_value in secret_build_args:
             name, value = name_value.split('=', 1)
             build_arguments.append(BuildArgument(name, value, True))
 
-    build_request = QueueBuildRequest(
-        image_name=image_name,
-        source_location=source_location,
-        build_parameters=build_parameters,
-        is_push_enabled=is_push_enabled,
-        timeout=timeout,
-        platform=platform,
-        build_arguments=build_arguments)
-
-    result = LongRunningOperation(cmd.cli_ctx)(client.queue(
-        build_request=build_request, resource_group_name=resource_group_name, registry_name=registry_name))
-
-    if isinstance(result, Build):
-        print("Successfully queued a build with build-id: {}. Starting to stream the logs...".format(result.build_id))
-        return acr_build_show_logs(cmd, client, registry_name, result.build_id, resource_group_name)
+    try:
+        build_request = QueueBuildRequest(
+            image_name=image_name,
+            source_location=source_location,
+            build_parameters=build_parameters,
+            is_push_enabled=is_push_enabled,
+            timeout=timeout,
+            platform=platform,
+            build_arguments=build_arguments)
+
+        result = LongRunningOperation(cmd.cli_ctx)(client.queue(
+            build_request=build_request, resource_group_name=resource_group_name, registry_name=registry_name))
+
+        print("Successfully queued a build with build-id: {}.".format(result.build_id))
+
+        if show_logs == 'true':
+            print("Starting to stream the logs...")
+            return acr_build_show_logs(cmd, client, registry_name, result.build_id, resource_group_name)
+    except Exception as err:
+        raise CLIError(err)
+
+
+def _check_remote_source_code(source_location):
+
+    # NOTE: acr_build doesn't support git repo url
+    if source_location.lower().startswith("git://"):
+        return source_location
+
+
+    response = requests.head(source_location)
+    if response.status_code < 400:
+        return source_location
+    else:
+        raise CLIError("'{}' doesn't exist.".format(source_location))
+
+
+def _check_image_name(image_name):
+    
+    # referenc: https://github.com/docker/distribution/tree/master/reference
+
+    if image_name == "":
+        raise CLIError("'--tag' value should not be empty.")
+    
+    tokens = image_name.split(':')
+    if(len(tokens) > 2):
+        raise CLIError("'--tag' value should be repository-name and optionally a tag in the 'repository-name:tag' format")
+
+    # check repository-name    
+    repository_name = tokens[0]
+    if len(repository_name) > 255:
+        raise CLIError("The repository-name of '--tag' value should be no more than 255 characters.")
     else:
-        # Maybe just raise error?
-        return result
+        if re.match(r"^[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?(?:(?:/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+)?$", repository_name) is None:
```

Comment:
Maybe a TODO in VSTS for this, we should pull in their tests and compare the differences. E.g., should we do the full grammar check. For example `validname@invaliddigest:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff` should fail but passes here. Don't know how much we care about it being perfect/consistent.

Reply:
Just avoid client side validation and let the server side validate. You'll run into maintenance and inconsistency headache easily with client validation.

Reply:
@djyou this isn't really something a server can validate...

Reply:
@djyou idk, for most of these I think it makes sense to have the client-side validation. For this one, it would be a bummer to send the request all the way down the entire pipeline just to have it fail, pretty bad user experience; we'd bill them for it; we'd hurt our resources. For the others we pretty much have to do them.

Reply:
A preferred pattern is to avoid client side validation because if tomorrow there is a change on the server side we will have to update the client side too. The even worse part is we have to ask users to update their client to match whatever the new server side validation does.

I suppose server side would have to validate image name anyway. We probably have more traffic from various SDKs directly or third party apps than from our own clients (i.e., CLI/Portal), plus there is no way we can prevent incoming garbage data. I agree with your concerns but compared with the potential maintenance effort, the real overhead on the user experience is no more than a few seconds. We can start billing after all validations are passed to avoid the billing issue.

That said, in some rare case client side validation can make a real difference so it's just a trade-off you folks can decide.